From d635f3a37f86bd99909f305184a483e6e51e0eae Mon Sep 17 00:00:00 2001
From: Siddharth Pandey <siddharth.pandey@vvdntech.com>
Date: Sat, 23 Oct 2021 22:37:08 +0530
Subject: [PATCH] ipq4019: Single PHY VLAN Fix.

WIFI-4226-AP HFCL: clinet can't access internet with SSID VLAN

cause: Ingress packets getting double VLAN tagged.

Fixed: Added patch for edma driver.

Signed-off-by: Siddharth Pandey <siddharth.pandey@vvdntech.com>
---
 .../A01-ipq4019-single_phy_vlan_fix.patch          | 28 ++++++++++++++++++++++
 1 file changed, 28 insertions(+)
 create mode 100644 target/linux/ipq40xx/patches-4.14/A01-ipq4019-single_phy_vlan_fix.patch

diff --git a/target/linux/ipq40xx/patches-4.14/A01-ipq4019-single_phy_vlan_fix.patch b/target/linux/ipq40xx/patches-4.14/A01-ipq4019-single_phy_vlan_fix.patch
new file mode 100644
index 0000000..8b73d28
--- /dev/null
+++ b/target/linux/ipq40xx/patches-4.14/A01-ipq4019-single_phy_vlan_fix.patch
@@ -0,0 +1,28 @@
+diff --git a/drivers/net/ethernet/qualcomm/essedma/edma.c b/drivers/net/ethernet/qualcomm/essedma/edma.c
+index 4c55224..ed6260b 100644
+--- a/drivers/net/ethernet/qualcomm/essedma/edma.c
++++ b/drivers/net/ethernet/qualcomm/essedma/edma.c
+@@ -164,10 +164,8 @@ static void edma_configure_rx(struct edma_common_info *edma_cinfo)
+ 	/* Set Rx FIFO threshold to start to DMA data to host */
+ 	rxq_ctrl_data = EDMA_FIFO_THRESH_128_BYTE;
+ 
+-	if (!edma_cinfo->is_single_phy) {
+ 		/* Set RX remove vlan bit */
+ 		rxq_ctrl_data |= EDMA_RXQ_CTRL_RMV_VLAN;
+-	}
+ 
+ 	edma_write_reg(EDMA_REG_RXQ_CTRL, rxq_ctrl_data);
+ }
+@@ -1413,12 +1411,10 @@ netdev_tx_t edma_xmit(struct sk_buff *skb,
+ 	}
+ 
+ 	/* Check and mark VLAN tag offload */
+-	if (!adapter->edma_cinfo->is_single_phy) {
+ 		if (unlikely(skb_vlan_tag_present(skb)))
+ 			flags_transmit |= EDMA_VLAN_TX_TAG_INSERT_FLAG;
+ 		else if (adapter->default_vlan_tag)
+ 			flags_transmit |= EDMA_VLAN_TX_TAG_INSERT_DEFAULT_FLAG;
+-	}
+ 
+ 	/* Check and mark checksum offload */
+ 	if (likely(skb->ip_summed == CHECKSUM_PARTIAL))
-- 
2.7.4

